compiler/src/iree/compiler/Codegen/Common/ConfigTrackingCanonicalizer.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/Passes.h"
+#include "iree/compiler/Codegen/Common/Transforms.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-codegen-config-tracking-canonicalizer"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_CONFIGTRACKINGCANONICALIZERPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+
+static Operation *skipCastsDefiningOp(Value v) {
+  auto producer = v.getDefiningOp();
+  while (auto castProducer = dyn_cast<tensor::CastOp>(producer)) {
+    producer = castProducer.getSource().getDefiningOp();
+  }
+  return producer;
+}
+
+void ConfigTrackingListener::notifyOperationReplaced(Operation *op,
```

Comment:
Don't you also have to listen for in-place modifications?

Reply:
In-place modifications don't change the set of attributes right? Unless someone removed discardable attrs on purpose, in which case I'm not sure how we'd track that.

Reply:
they can

Reply:
but that's probably much less likely

Reply:
I see, we could use `startOpModification` and `finalizeOpModification` and track the lowering config through that. We can add that once someone decides to be "principled" with another pattern.

Reply:
Note that the pattern rewrite drivers require you to go through either these or `.modifyOpInPlace` (which ends up the those two functions), so if we don't already do that, it's a bug.

To track these, you could memorize the attribute state in the `notifyPatternBegin` callback and then check in `notifyOperationModified`.